File: ScopeFFT.ts
/*
    Copyright (C) Paul Falstad and Iain Sharp

    This file is part of CircuitJS1.

    CircuitJS1 is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 2 of the License, or
    (at your option) any later version.

    CircuitJS1 is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CircuitJS1.  If not, see <http://www.gnu.org/licenses/>.
*/

import { Graphics } from "./Graphics";
import { CircuitElm } from "./CircuitElm";
import { Locale } from "./Locale";
import { FFT } from "./FFT";
import { UNITS_V, UNITS_A } from "./ScopeConstants";
import type { Scope } from "./Scope";
import type { ScopePlot } from "./ScopePlot";

export class ScopeFFT {
    scope: Scope;
    enabled: boolean = false;
    logSpectrum: boolean = false;
    showPhaseAngle: boolean = false;
    fftMaxMagnitude: number = 0;
    fftReal: number[] | null = null;
    fftImag: number[] | null = null;
    private fft: FFT | null = null;

    constructor(scope: Scope) {
        this.scope = scope;
    }

    show(b: boolean): void {
        this.enabled = b;
        if (!this.enabled)
            this.fft = null;
    }

    drawVerticalGridLines(g: Graphics): void {
        let prevEnd = 0;
        const divs = 20;
        const maxFrequency = 1 / (this.scope.sim.maxTimeStep * this.scope.speed * divs * 2);
        for (let i = 0; i < divs; i++) {
            const x = Math.floor(this.scope.rect.width * i / divs);
            if (x < prevEnd) continue;
            const s = "" + Math.round(i * maxFrequency) + "Hz";
            const sWidth = Math.ceil(g.context.measureText(s).width);
            prevEnd = x + sWidth + 4;
            if (i > 0) {
                g.setColor("#880000");
                g.drawLine(x, 0, x, this.scope.rect.height);
            }
            g.setColor("#FF0000");
            g.drawString(s, x + 2, this.scope.rect.height);
        }
    }

    draw(g: Graphics): void {
        if (this.fft === null || this.fft.getSize() !== this.scope.scopePointCount)
            this.fft = new FFT(this.scope.scopePointCount);
        const real = new Array<number>(this.scope.scopePointCount).fill(0);
        const imag = new Array<number>(this.scope.scopePointCount).fill(0);
        const plot: ScopePlot = (this.scope.visiblePlots.length === 0)
            ? this.scope.plots[0]
            : this.scope.visiblePlots[0];
        const maxV = plot.maxValues;
        const minV = plot.minValues;
        const ptr = plot.ptr;
        for (let i = 0; i < this.scope.scopePointCount; i++) {
            const ii = (ptr - i + this.scope.scopePointCount) & (this.scope.scopePointCount - 1);
            // average max and min to prevent DC spike from masking rest of spectrum
            real[i] = 0.5 * (maxV[ii] + minV[ii]);
            imag[i] = 0;
        }
        this.fft.fft(real, imag, true);
        let maxM = 1e-8;
        for (let i = 0; i < this.scope.scopePointCount / 2; i++) {
            const m = this.fft.magnitude(real[i], imag[i]);
            if (m > maxM)
                maxM = m;
        }
        this.fftMaxMagnitude = maxM;
        this.fftReal = real;
        this.fftImag = imag;
        let prevX = 0;
        g.setColor("#FF0000");
        if (!this.logSpectrum) {
            let prevHeight = 0;
            const y = (this.scope.rect.height - 1) - 12;
            for (let i = 0; i < this.scope.scopePointCount / 2; i++) {
                const x = Math.floor(2 * i * this.scope.rect.width / this.scope.scopePointCount);
                const magnitude = this.fft.magnitude(real[i], imag[i]);
                const height = Math.floor((magnitude * y) / maxM);
                if (x !== prevX)
                    g.drawLine(prevX, y - prevHeight, x, y - height);
                prevHeight = height;
                prevX = x;
            }
        } else {
            const dbRange = 80;
            const topMargin = 5;
            const bottomMargin = 12;
            const plotHeight = this.scope.rect.height - topMargin - bottomMargin;
            const pixelsPerDb = plotHeight / dbRange;
            let prevY = 0;
            for (let db = -20; db >= -80; db -= 20) {
                const y = topMargin + Math.floor(-db * pixelsPerDb);
                if (y < 0 || y >= this.scope.rect.height)
                    continue;
                g.setColor("#880000");
                g.drawLine(0, y, this.scope.rect.width, y);
                g.setColor("#FF0000");
                g.drawString(db + " dB", 2, y - 2);
            }
            g.setColor("#FF0000");
            for (let i = 0; i < this.scope.scopePointCount / 2; i++) {
                const x = Math.floor(2 * i * this.scope.rect.width / this.scope.scopePointCount);
                const magnitude = this.fft.magnitude(real[i], imag[i]);
                let db = 20 * Math.log(magnitude / maxM) / Math.log(10);
                if (db < -dbRange)
                    db = -dbRange;
                const y = topMargin + Math.floor(-db * pixelsPerDb);
                if (x !== prevX)
                    g.drawLine(prevX, prevY, x, y);
                prevY = y;
                prevX = x;
            }
        }
    }

    drawPhaseAngle(g: Graphics): void {
        let vPlot: ScopePlot | null = null;
        let iPlot: ScopePlot | null = null;
        for (const p of this.scope.visiblePlots) {
            if (p.units === UNITS_V) {
                if (vPlot !== null) return;
                vPlot = p;
            } else if (p.units === UNITS_A) {
                if (iPlot !== null) return;
                iPlot = p;
            } else
                return;
        }
        if (vPlot === null || iPlot === null)
            return;
        if (this.fft === null || this.fft.getSize() !== this.scope.scopePointCount)
            this.fft = new FFT(this.scope.scopePointCount);
        const vReal = new Array<number>(this.scope.scopePointCount).fill(0);
        const vImag = new Array<number>(this.scope.scopePointCount).fill(0);
        const iReal = new Array<number>(this.scope.scopePointCount).fill(0);
        const iImag = new Array<number>(this.scope.scopePointCount).fill(0);
        const ipa = this.scope.displayStartIndex(vPlot, this.scope.rect.width);
        const validCount = this.scope.validDataCount(vPlot, ipa, this.scope.rect.width);
        for (let i = 0; i < validCount; i++) {
            const ip = (i + ipa) & (this.scope.scopePointCount - 1);
            vReal[i] = 0.5 * (vPlot.maxValues[ip] + vPlot.minValues[ip]);
            iReal[i] = 0.5 * (iPlot.maxValues[ip] + iPlot.minValues[ip]);
        }
        this.fft.fft(vReal, vImag, true);
        this.fft.fft(iReal, iImag, true);
        let fund = 1;
        let maxM = 0;
        for (let i = 1; i < this.scope.scopePointCount / 2; i++) {
            const m = this.fft.magnitude(vReal[i], vImag[i]);
            if (m > maxM) { maxM = m; fund = i; }
        }
        if (maxM < 1e-8)
            return;
        const angleV = Math.atan2(vImag[fund], vReal[fund]);
        const angleI = Math.atan2(iImag[fund], iReal[fund]);
        let angle = (angleV - angleI) * 180 / Math.PI;
        while (angle > 180) angle -= 360;
        while (angle < -180) angle += 360;
        this.scope.drawInfoText(g, Locale.LS("Phase angle: ") + CircuitElm.showFormat.format(angle) + "°");
    }

    addCursorInfo(info: string[], ct: number, mouseCursorX: number): number {
        const maxFrequency = 1 / (this.scope.sim.maxTimeStep * this.scope.speed * 2);
        info[ct++] = CircuitElm.getUnitText(maxFrequency * (mouseCursorX - this.scope.rect.x) / this.scope.rect.width, "Hz");
        if (this.fft !== null && this.fftReal !== null && this.fftMaxMagnitude > 0) {
            const fftIndex = Math.floor((mouseCursorX - this.scope.rect.x) * this.scope.scopePointCount / (2 * this.scope.rect.width));
            if (fftIndex >= 0 && fftIndex < this.scope.scopePointCount / 2) {
                const mag = this.fft.magnitude(this.fftReal[fftIndex], this.fftImag![fftIndex]);
                const db = 20 * Math.log(mag / this.fftMaxMagnitude) / Math.log(10);
                info[ct++] = Math.round(db) + " dB";
            }
        }
        return ct;
    }
}

File: src/com/lushprojects/circuitjs1/client/ScopeFFT.java
package com.lushprojects.circuitjs1.client;

import com.lushprojects.circuitjs1.client.util.Locale;

class ScopeFFT {
    Scope scope;
    boolean enabled;
    boolean logSpectrum;
    boolean showPhaseAngle;
    double fftMaxMagnitude;
    double[] fftReal, fftImag;
    private FFT fft;

    ScopeFFT(Scope scope) {
	this.scope = scope;
    }

    void show(boolean b) {
	enabled = b;
	if (!enabled)
	    fft = null;
    }

    void drawVerticalGridLines(Graphics g) {
	int prevEnd = 0;
	int divs = 20;
	double maxFrequency = 1 / (scope.sim.maxTimeStep * scope.speed * divs * 2);
	for (int i = 0; i < divs; i++) {
	    int x = scope.rect.width * i / divs;
	    if (x < prevEnd) continue;
	    String s = ((int) Math.round(i * maxFrequency)) + "Hz";
	    int sWidth = (int) Math.ceil(g.context.measureText(s).getWidth());
	    prevEnd = x + sWidth + 4;
	    if (i > 0) {
		g.setColor("#880000");
		g.drawLine(x, 0, x, scope.rect.height);
	    }
	    g.setColor("#FF0000");
	    g.drawString(s, x + 2, scope.rect.height);
	}
    }

    void draw(Graphics g) {
	if (fft == null || fft.getSize() != scope.scopePointCount)
	    fft = new FFT(scope.scopePointCount);
	double[] real = new double[scope.scopePointCount];
	double[] imag = new double[scope.scopePointCount];
	ScopePlot plot = (scope.visiblePlots.size() == 0) ? scope.plots.firstElement() : scope.visiblePlots.firstElement();
	double maxV[] = plot.maxValues;
	double minV[] = plot.minValues;
	int ptr = plot.ptr;
	for (int i = 0; i < scope.scopePointCount; i++) {
	    int ii = (ptr - i + scope.scopePointCount) & (scope.scopePointCount - 1);
	    // average max and min to prevent DC spike from masking rest of spectrum
	    real[i] = .5 * (maxV[ii] + minV[ii]);
	    imag[i] = 0;
	}
	fft.fft(real, imag, true);
	double maxM = 1e-8;
	for (int i = 0; i < scope.scopePointCount / 2; i++) {
	    double m = fft.magnitude(real[i], imag[i]);
	    if (m > maxM)
		maxM = m;
	}
	fftMaxMagnitude = maxM;
	fftReal = real;
	fftImag = imag;
	int prevX = 0;
	g.setColor("#FF0000");
	if (!logSpectrum) {
	    int prevHeight = 0;
	    int y = (scope.rect.height - 1) - 12;
	    for (int i = 0; i < scope.scopePointCount / 2; i++) {
		int x = 2 * i * scope.rect.width / scope.scopePointCount;
		double magnitude = fft.magnitude(real[i], imag[i]);
		int height = (int) ((magnitude * y) / maxM);
		if (x != prevX)
		    g.drawLine(prevX, y - prevHeight, x, y - height);
		prevHeight = height;
		prevX = x;
	    }
	} else {
	    double dbRange = 80;
	    int topMargin = 5;
	    int bottomMargin = 12;
	    int plotHeight = scope.rect.height - topMargin - bottomMargin;
	    double pixelsPerDb = plotHeight / dbRange;
	    int prevY = 0;
	    for (int db = -20; db >= -80; db -= 20) {
		int y = topMargin + (int) (-db * pixelsPerDb);
		if (y < 0 || y >= scope.rect.height)
		    continue;
		g.setColor("#880000");
		g.drawLine(0, y, scope.rect.width, y);
		g.setColor("#FF0000");
		g.drawString(db + " dB", 2, y - 2);
	    }
	    g.setColor("#FF0000");
	    for (int i = 0; i < scope.scopePointCount / 2; i++) {
		int x = 2 * i * scope.rect.width / scope.scopePointCount;
		double magnitude = fft.magnitude(real[i], imag[i]);
		double db = 20 * Math.log(magnitude / maxM) / Math.log(10);
		if (db < -dbRange)
		    db = -dbRange;
		int y = topMargin + (int) (-db * pixelsPerDb);
		if (x != prevX)
		    g.drawLine(prevX, prevY, x, y);
		prevY = y;
		prevX = x;
	    }
	}
    }

    void drawPhaseAngle(Graphics g) {
	ScopePlot vPlot = null, iPlot = null;
	for (ScopePlot p : scope.visiblePlots) {
	    if (p.units == Scope.UNITS_V) {
		if (vPlot != null) return;
		vPlot = p;
	    } else if (p.units == Scope.UNITS_A) {
		if (iPlot != null) return;
		iPlot = p;
	    } else
		return;
	}
	if (vPlot == null || iPlot == null)
	    return;
	if (fft == null || fft.getSize() != scope.scopePointCount)
	    fft = new FFT(scope.scopePointCount);
	double[] vReal = new double[scope.scopePointCount];
	double[] vImag = new double[scope.scopePointCount];
	double[] iReal = new double[scope.scopePointCount];
	double[] iImag = new double[scope.scopePointCount];
	int ipa = scope.displayStartIndex(vPlot, scope.rect.width);
	int validCount = scope.validDataCount(vPlot, ipa, scope.rect.width);
	for (int i = 0; i < validCount; i++) {
	    int ip = (i + ipa) & (scope.scopePointCount - 1);
	    vReal[i] = .5 * (vPlot.maxValues[ip] + vPlot.minValues[ip]);
	    iReal[i] = .5 * (iPlot.maxValues[ip] + iPlot.minValues[ip]);
	}
	fft.fft(vReal, vImag, true);
	fft.fft(iReal, iImag, true);
	int fund = 1;
	double maxM = 0;
	for (int i = 1; i < scope.scopePointCount / 2; i++) {
	    double m = fft.magnitude(vReal[i], vImag[i]);
	    if (m > maxM) { maxM = m; fund = i; }
	}
	if (maxM < 1e-8)
	    return;
	double angleV = Math.atan2(vImag[fund], vReal[fund]);
	double angleI = Math.atan2(iImag[fund], iReal[fund]);
	double angle = (angleV - angleI) * 180 / Math.PI;
	while (angle > 180) angle -= 360;
	while (angle < -180) angle += 360;
	scope.drawInfoText(g, Locale.LS("Phase angle: ") + CircuitElm.showFormat.format(angle) + "°");
    }

    int addCursorInfo(String[] info, int ct, int mouseCursorX) {
	double maxFrequency = 1 / (scope.sim.maxTimeStep * scope.speed * 2);
	info[ct++] = CircuitElm.getUnitText(maxFrequency * (mouseCursorX - scope.rect.x) / scope.rect.width, "Hz");
	if (fft != null && fftReal != null && fftMaxMagnitude > 0) {
	    int fftIndex = (mouseCursorX - scope.rect.x) * scope.scopePointCount / (2 * scope.rect.width);
	    if (fftIndex >= 0 && fftIndex < scope.scopePointCount / 2) {
		double mag = fft.magnitude(fftReal[fftIndex], fftImag[fftIndex]);
		double db = 20 * Math.log(mag / fftMaxMagnitude) / Math.log(10);
		info[ct++] = Math.round(db) + " dB";
	    }
	}
	return ct;
    }
}
